### PR TITLE
refactor authentication and authorization

### PR DESCRIPTION
**Pull request by BlaiseJavan**
#### What does this PR do?

- Refactor authentication and authorization

#### Description of Task to be completed?

- Name all the task completed in reference to the one listed in the Jira description

  -  create roles

  -  access controller 

#### How should this be manually tested?

- Clone the project through your terminal(VScode terminal)

      git clone https://github.com/Volve-gitore/quickss-backend.git

- Go to the project and Open the project

       cd quickss-backend && cd .

- Open the feature through its branch

       - git checkout refactor-auth

- Install all the dependencies

       npm i || yarn install

#### Any background context you want to provide?

- N/A

#### What are the relevant Jira story?

- N/A

#### Screenshots (if appropriate)

- N/A

### CODE REVIEW

File: src/controllers/role.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+   *
+   * @desc Create new role
+   * @route POST /api/role/
+   */
+  static async createRole(req, res) {
+    try {
+      const { name } = req.body;
+      const roleExist = await Role.findOne({ where: { name: name } });
+      if (roleExist) {
+        return res.status(400).send({
```
Comment (Emile-Nsengimana) [inline src/controllers/role.js:17]: status code should be `409` for conflict

File: src/controllers/role.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  static async getRoles(req, res) {
+    try {
+      const roles = await Role.findAll();
+      if (roles.length === 0)
+        return res.status(404).send({
+          error: 'no role found',
+        });
+
+      return res.status(200).send({
+        items: roles,
```
Comment (Emile-Nsengimana) [inline src/controllers/role.js:50]: better to have `roles` instead of `items` for convenience

File: src/controllers/type.js [hunk 3]
```diff
@@ -11,16 +11,12 @@ class TypeManager {
   static async createType(req, res) {
     try {
       const { name, clientId } = req.body;
-      console.log('hh')
-
       const typeExist = await Type.findOne({ where: { name: name, clientId: clientId } });
       if (typeExist) {
         return res.status(400).send({
```
Comment (Emile-Nsengimana) [inline src/controllers/type.js:16]: change the status code from `400` to `409` (conflict)

File: src/roles.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+exports.roles = (function() {
+
+ac.grant("customer")
+  .readAny("hotelResto")    
+
+ac.grant("client")
+  .extend("customer")
+
+ .createOwn("hotelResto")
+ .updateOwn("hotelResto")
```
Comment (Emile-Nsengimana) [inline src/roles.js:13]: do we still use `hotelResto` or `clients`?

File: src/routes/client.js [hunk 5]
```diff
@@ -4,11 +4,12 @@ import { multerUploads } from '../middlewares/multer';
...
 import { clientSchema } from '../validations/schema/client';
+import auth from '../middlewares/auth'
 
 const router = express.Router();
 
 router.use('*', cloudinaryConfig);
-router.post('/', multerUploads, validator(clientSchema), clientController.registerClient);
-router.get('/', clientController.getClients);
+router.post('/', auth.grantAccess('createOwn', 'hotelResto'), multerUploads, validator(clientSchema), clientController.registerClient);
+router.get('/', auth.grantAccess('readAny', 'hotelResto'), clientController.getClients);
```
Comment (Emile-Nsengimana) [inline src/routes/client.js:13]: please make sure that all references of `hotelResto` are replaced with `clients`

File: yarn.lock [hunk 6]
```diff
@@ -4782,6 +4789,11 @@ normalize-url@^4.1.0:
...
   integrity sha512-2s47yzUxdexf1OhyRi4Em83iQk0aPvwTddtFz4hnSSw9dCEsLEGf6SwIO8ss/19S9iBb5sJaOuTvTGDeZI00BQ==
 
+notation@^1.3.6:
+  version "1.3.6"
+  resolved "https://registry.yarnpkg.com/notation/-/notation-1.3.6.tgz#bc87b88d1f1159e2931e7f9317a3020313790321"
+  integrity sha512-DIuJmrP/Gg1DcXKaApsqcjsJD6jEccqKSfmU3BUx/f1GHsMiTJh70cERwYc64tOmTRTARCeMwkqNNzjh3AHhiw==
+
 npmlog@^4.1.2:
   version "4.1.2"
   resolved "https://registry.yarnpkg.com/npmlog/-/npmlog-4.1.2.tgz#08a7f2a8bf734604779a9efa4ad5cc717abb954b"
```
Comment (Emile-Nsengimana) [inline yarn.lock:4799]: this file can be added on `.gitIgnore`

File: src/middlewares/auth.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+// Add this to the top of the file
+import model from '../db/models';
+import UserService from '../services/user';
+
+const jwt = require('jsonwebtoken');
```
Comment (Emile-Nsengimana) [inline src/middlewares/auth.js:5]: Can we stick to `ES6>` syntax

**Review by Emile-Nsengimana — CHANGES_REQUESTED** [review]

File: src/roles.js (hunk 4 above)
Comment (BlaiseJavan) [reply src/roles.js:13]: this is referring to a hotel and resto as an items not a roles.

File: src/routes/client.js (hunk 5 above)
Comment (BlaiseJavan) [reply src/routes/client.js:13]: same here.. we cant have a client as role and as action at the same time.. this hotelResto is it represents client on actions(create, read, update, delete).. but as role we have client.

**Review by Emile-Nsengimana — APPROVED** [review]

**Pull request merged**